org.eclipse.xtext/src/org/eclipse/xtext/formatting2/AbstractJavaFormatter.java

```diff
@@ -0,0 +1,180 @@
+/*******************************************************************************
+ * Copyright (c) 2020 itemis AG (http://www.itemis.eu) and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.xtext.formatting2;
+
+import java.util.Collections;
+
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.xtext.formatting2.regionaccess.IEObjectRegion;
+import org.eclipse.xtext.formatting2.regionaccess.IHiddenRegion;
+import org.eclipse.xtext.formatting2.regionaccess.ISemanticRegion;
+import org.eclipse.xtext.formatting2.regionaccess.ISemanticRegionFinder;
+import org.eclipse.xtext.formatting2.regionaccess.ISemanticRegionsFinder;
+import org.eclipse.xtext.resource.XtextResource;
+import org.eclipse.xtext.util.Exceptions;
+import org.eclipse.xtext.util.PolymorphicDispatcher;
+import org.eclipse.xtext.util.PolymorphicDispatcher.ErrorHandler;
+
+import com.google.common.annotations.Beta;
+
+/**
+ * <p>
+ * This is an abstract base class for language-specific formatters, intended to be extended with a java class.
+ * </p>
+ * <p>
+ * It is working very much like {@link AbstractFormatter2} but does support additional methods to simplify java
+ * formatter code.
+ * </p>
+ * <p>
+ * the "_format" methods are called by reflection. The name is important as well as there are exactly two arguments
+ * whereas the first one is of the element type you like to format and the second one of type
+ * {@link IFormattableDocument}. The methods should be protected.
+ * </p>
+ * <p>
+ * Example code for a java formatter:
+ * 
+ * <pre>
+ * import static org.xtext.example.mydsl.myDsl.MyDslPackage.Literals.*;
+ * 
+ * public class MyDslFormatter2 extends AbstractJavaFormatter {
+ * 	protected void _format(Model model, IFormattableDocument doc) {
+ * 		for (Parent parent : model.getParents())
+ * 			doc.format(parent);
+ * 	}
+ * 
+ * 	protected void _format(Parent parent, IFormattableDocument doc) {
+ * 		doc.prepend(regionFor(parent).keyword("parent"), it -> it.noSpace());
```

Comment:
please use a more descriptive name for `it` or replace by method reference `TypeOfIt::noSpace`